src/Test/ImplicitMethodsIntegrationTest.php

```diff
@@ -0,0 +1,226 @@
+ * @copyright Copyright (c) 2018 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive-router/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Router\Test;
+
+use Fig\Http\Message\RequestMethodInterface as RequestMethod;
+use Fig\Http\Message\StatusCodeInterface as StatusCode;
+use Generator;
+use PHPUnit\Framework\TestCase;
+use Prophecy\Argument;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Diactoros\Response;
+use Zend\Diactoros\ServerRequest;
+use Zend\Diactoros\Stream;
+use Zend\Expressive\Router\Middleware\ImplicitHeadMiddleware;
+use Zend\Expressive\Router\Middleware\ImplicitOptionsMiddleware;
+use Zend\Expressive\Router\Middleware\PathBasedRoutingMiddleware;
+use Zend\Expressive\Router\Route;
+use Zend\Expressive\Router\RouteResult;
+use Zend\Expressive\Router\RouterInterface;
+
+/**
+ * Base class for testing adapter integrations.
+ *
+ * Implementers of adapters should extend this class in their test suite,
+ * implementing the `getRouter()` method.
+ *
+ * This test class tests that the router correctly marshals the allowed methods
+ * for a match that matches the path, but not the request method.
+ */
+abstract class ImplicitMethodsIntegrationTest extends TestCase
+{
+    /**
+     * @return RouterInterface
+     */
+    abstract public function getRouter();
+
+    /**
+     * @return Generator
+     */
+    public function method()
+    {
+        yield RequestMethod::METHOD_HEAD => [
+            RequestMethod::METHOD_HEAD,
+            new ImplicitHeadMiddleware(
+                function () {
+                    return new Response();
+                },
+                function () {
+                    return new Stream('php://temp', 'rw');
+                }
+            ),
+        ];
+        yield RequestMethod::METHOD_OPTIONS => [
+            RequestMethod::METHOD_OPTIONS,
+            new ImplicitOptionsMiddleware(
+                function () {
+                    return new Response();
+                }
+            ),
+        ];
+    }
+
+    /**
+     * @dataProvider method
+     * @param string $method
+     */
+    public function testExplicitRequest($method, MiddlewareInterface $middleware)
+    {
+        $middleware1 = $this->prophesize(MiddlewareInterface::class)->reveal();
+        $middleware2 = $this->prophesize(MiddlewareInterface::class)->reveal();
+        $route1 = new Route('/api/v1/me', $middleware1, [$method]);
+        $route2 = new Route('/api/v1/me', $middleware2, [RequestMethod::METHOD_GET]);
+
+        $router = $this->getRouter();
+        $router->addRoute($route1);
+        $router->addRoute($route2);
+
+        $finalResponse = (new Response())->withHeader('foo-bar', 'baz');
+        $finalResponse->getBody()->write('FOO BAR BODY');
+
+        $finalHandler = $this->prophesize(RequestHandlerInterface::class);
```

Comment:
`RequestHandlerInterface` is also not defined, we should use `Webimpress\HttpMiddlewareCompatibility\HandlerInterface` here and in other places